### DISCUSSION TITLE
E_MISSING_MODEL_ATTRIBUTE with hasMany and belongsTo

### DISCUSSION THREAD

**Discussion by SedueRey** [adonisjs/core#3823] [Help] [answered]
Hello everyone,

Sorry if this is not the right way to ask but I don't know how to configure these relationships anymore. I have cards and turns in a n:m relationship, 1 card belongs to 1 turn and turns have many cards.

For the n:m relationships I have had no problem in the configuration but I keep getting the error when I try to create the relationship between cards and turns:

`Exception: E_MISSING_MODEL_ATTRIBUTE: "Turn.cards" expects "turnId" to exist on "Card" model, but is missing`

Also, in the database, after migration, both tables are binded in the right way.

Turn model:

```
export default class Turn extends BaseModel {
  @beforeCreate()
  public static async createUUID(model: Turn) {
    if (!model.id) {
      model.id = await uuid()
    }
  }

  @column({ isPrimary: true })
  public id: string

  @hasMany(() => Card)
  public cards: HasMany<typeof Card>
}

```

Card model:
```
export default class Card extends BaseModel {
  @beforeCreate()
  public static async createUUID(model: Card) {
    if (!model.id) {
      model.id = await uuid()
    }
  }

  @column({ isPrimary: true })
  public id: string

  @column()
  public price: number

  @column.dateTime({ autoCreate: true, serializeAs: null })
  public createdAt: DateTime

  @column.dateTime({ autoCreate: true, autoUpdate: true, serializeAs: null })
  public updatedAt: DateTime

  @manyToMany(() => Language, {
    pivotColumns: ['title'],
    pivotTimestamps: true,
  })
  public languages: ManyToMany<typeof Language>

  @manyToMany(() => Cardtype)
  public cardtypes: ManyToMany<typeof Cardtype>

  @belongsTo(() => Player)
  public player: BelongsTo<typeof Player>

  @belongsTo(() => Turn)
  public turn: BelongsTo<typeof Turn>
}

```

Migration defining relationship:
```
protected tableName = 'turns'

public async up() {
  this.schema.createTable(this.tableName, (table) => {
    table.uuid('id').primary()
    table.dateTime('startdate').notNullable()
    table.dateTime('enddate').notNullable()

    table.timestamps(true, true)
  })
  this.schema.alterTable('cards', (table) => {
    table.uuid('turn_id').references('turns.id')
  })
}
```
When I try to asign a Turn to a Card or viceversa in, for example, a seeder, i get the error. Example seeder code would be:
```
const createdTurn = await Turn.create({
  id: turnid, // an v4 uuid
  startdate: DateTime.now().plus({ weeks: i }),
  enddate: DateTime.now().plus({ weeks: i + 1 }),
})
const card = await Card.query({ client: trx }).where('id', 'the-card-uuid').first()
await createdTurn.related('cards').save(card)
```

`await card.related('turn').associate(createdTurn)` also raises same error.

I am quite lost and do not see what could be causing the error. I have tried what other people have asked but have not been able to get it to work for me either.

Thanks in advance

Best regards,

**Comment by Barbapapazes** [comment]
Hello, have you read pages about these relations?

- [Has one](https://docs.adonisjs.com/reference/orm/relations/has-one)
- [Has many](https://docs.adonisjs.com/reference/orm/relations/has-many)
- [Belongs to](https://docs.adonisjs.com/reference/orm/relations/belongs-to)
- [Many to many](https://docs.adonisjs.com/reference/orm/relations/many-to-many)
- [Has many through](https://docs.adonisjs.com/reference/orm/relations/has-many-through)

And can you provide a schema model using [DrawSQL](https://drawsql.app/) to understand the purpose?

**Reply by Barbapapazes** [reply]
But as I can understand, a `turnId` key is missing in the card model.

**Comment by SedueRey** [comment]
Thanks for answering, @Barbapapazes 

Summarizing the relationship, the diagram would look like this https://drawsql.app/s-28/diagrams/minimal/embed

1 turn -> many cards (the rest of the relationships (many to many) are not relevant to this error and, also, are working great)

I have read the documentation and I have called the attributes (as far as I have seen and if I am not mistaken) in the same way as they are called in the examples, right?

If i do some console.log with `.$getRelation` I also get the right values

**Comment by SedueRey** [comment, marked as answer]
Oh... I figured out what was going on. It doesn't say it in the documentation, but **you need to also add the corresponding attribute to the relationship**, not just the relationship itself.

As in the migration it is done together I hadn't thought it was necessary.

So it has been fixed with this code in the "n" relationship, the one with `BelongsTo` defined (I leave it here in case someone can use it):

````
  @column()
  public turnId: string
```

Should I make a PR to docs with this information?

**Reply by Barbapapazes** [reply]
In your case, that's just some SQL logic, not related to Adonis! 

And foreignKey is explained [here in documentation](https://docs.adonisjs.com/reference/orm/relations/has-one#foreignkey)!
